src/state_mgmt/state_machine.gd

```diff
@@ -3,129 +3,136 @@ class_name FrayStateMachine
 extends Node
 ## General purpose hierarchical state machine
 ##
-## This class wraps around the [FrayRootState] and uses the [SceneTree] to
+## This class wraps around the [FrayCompoundState] and uses the [SceneTree] to
 ## process state nodes.
 
 ## Emitted when the current state within the root changes.
 signal state_changed(from: StringName, to: StringName)
 
-enum AdvanceMode{
-	IDLE, ## Advance during the idle process
-	PHYSICS, ## Advance during the physics process
-	MANUAL, ## Advance manually
+enum AdvanceMode {
+	IDLE,  ## Advance during the idle process
+	PHYSICS,  ## Advance during the physics process
+	MANUAL,  ## Advance manually
 }
 
-## If true the combat state machine will be processing.
+## If true the [FrayStateMachine] will be processing.
 @export var active: bool = false
 
 ## Determines the process during which the state machine can advance.
-## Advancing only relates to transitions. 
+## Advancing only relates to transitions.
 ## If the state machine is active then the current state is still processed
 ## during both idle and physics frames regardless of advance mode.
 @export var advance_mode: AdvanceMode = AdvanceMode.IDLE
 
 ## The root of this state machine.
-var root: FrayRootState:
-	get: return _root
-	set(value): _set_root(value)
-
-## The state machine's current state.
-## Updating this value is the equivalent to calling [code]goto(state)[/code].
-var current_state: StringName = "":
-	get: 
-		if _ERR_ROOT_NOT_SET("Failed to get current state"): return ""
-		return _root.current_state
+var root: FrayCompoundState:
+	get:
+		return root
 	set(value):
-		if _ERR_ROOT_NOT_SET("Failed to set current state"): return
+		if root != null and root.transitioned.is_connected(_on_RootState_transitioned):
+			root.transitioned.disconnect(_on_RootState_transitioned)
 
-		root.current_state = value
+		root = value
+		root._enter_impl({})
+		root._init_state_impl()
```

Comment:
_init_state_impl shouldn't be called in the root setter. Only when building the states in add_state.

Reply:
My mistake, the root setter is for applying the CompoundState to the StateMachine.